common/src/main/cpp/extensions/openxr_fb_composition_layer_settings_extension_wrapper.cpp

```diff
@@ -0,0 +1,168 @@
+#include <godot_cpp/variant/utility_functions.hpp>
+
+using namespace godot;
+
+static const char *SUPERSAMPLING_MODE_PROPERTY_NAME = "XR_FB_composition_layer_settings/supersampling_mode";
+static const char *SHARPENING_MODE_PROPERTY_NAME = "XR_FB_composition_layer_settings/sharpening_mode";
+
+OpenXRFbCompositionLayerSettingsExtensionWrapper *OpenXRFbCompositionLayerSettingsExtensionWrapper::singleton = nullptr;
+
+OpenXRFbCompositionLayerSettingsExtensionWrapper *OpenXRFbCompositionLayerSettingsExtensionWrapper::get_singleton() {
+	if (singleton == nullptr) {
+		singleton = memnew(OpenXRFbCompositionLayerSettingsExtensionWrapper());
+	}
+	return singleton;
+}
+
+OpenXRFbCompositionLayerSettingsExtensionWrapper::OpenXRFbCompositionLayerSettingsExtensionWrapper() :
+		OpenXRExtensionWrapperExtension() {
+	ERR_FAIL_COND_MSG(singleton != nullptr, "An OpenXRFbCompositionLayerSettingsExtensionWrapper singleton already exists.");
+
+	request_extensions[XR_FB_COMPOSITION_LAYER_SETTINGS_EXTENSION_NAME] = &fb_composition_layer_settings;
+	singleton = this;
+}
+
+OpenXRFbCompositionLayerSettingsExtensionWrapper::~OpenXRFbCompositionLayerSettingsExtensionWrapper() {
+	cleanup();
+}
+
+void OpenXRFbCompositionLayerSettingsExtensionWrapper::_bind_methods() {
+	BIND_ENUM_CONSTANT(SUPERSAMPLING_MODE_DISABLED);
+	BIND_ENUM_CONSTANT(SUPERSAMPLING_MODE_NORMAL);
+	BIND_ENUM_CONSTANT(SUPERSAMPLING_MODE_QUALITY);
+
+	BIND_ENUM_CONSTANT(SHARPENING_MODE_DISABLED);
+	BIND_ENUM_CONSTANT(SHARPENING_MODE_NORMAL);
+	BIND_ENUM_CONSTANT(SHARPENING_MODE_QUALITY);
+}
+
+void OpenXRFbCompositionLayerSettingsExtensionWrapper::cleanup() {
+	fb_composition_layer_settings = false;
+}
+
+Dictionary OpenXRFbCompositionLayerSettingsExtensionWrapper::_get_requested_extensions() {
+	Dictionary result;
+	for (auto ext : request_extensions) {
+		uint64_t value = reinterpret_cast<uint64_t>(ext.value);
+		result[ext.key] = (Variant)value;
+	}
+	return result;
+}
+
+uint64_t OpenXRFbCompositionLayerSettingsExtensionWrapper::_set_viewport_composition_layer_and_get_next_pointer(const void *p_layer, const Dictionary &p_property_values, void *p_next_pointer) {
+	if (!fb_composition_layer_settings) {
+		return reinterpret_cast<uint64_t>(p_next_pointer);
+	}
+
+	const XrCompositionLayerBaseHeader *layer = reinterpret_cast<const XrCompositionLayerBaseHeader *>(p_layer);
+
+	if (!layer_structs.has(layer)) {
+		layer_structs[layer] = {
+			XR_TYPE_COMPOSITION_LAYER_SETTINGS_FB, // type
+			p_next_pointer, // next
+			0, // layerFlags
+		};
+	}
+
+	XrCompositionLayerSettingsFB *settings = layer_structs.getptr(layer);
+
+	switch ((SupersamplingMode)(int)p_property_values.get(SUPERSAMPLING_MODE_PROPERTY_NAME, SUPERSAMPLING_MODE_DISABLED)) {
+		case SUPERSAMPLING_MODE_DISABLED: {
+			settings->layerFlags = (settings->layerFlags & ~XR_COMPOSITION_LAYER_SETTINGS_NORMAL_SUPER_SAMPLING_BIT_FB) & ~XR_COMPOSITION_LAYER_SETTINGS_QUALITY_SUPER_SAMPLING_BIT_FB;
```

Comment:
These bitwise operations work, however, we could make all this code much simpler by clearing `settings->layerFlags` before the switch statements. Then these could all just be additive.

So, something like:

```
settings->layerFlags = 0;

switch ((SupersamplingMode)(int)p_property_values.get(SUPERSAMPLING_MODE_PROPERTY_NAME, SUPERSAMPLING_MODE_DISABLED)) {
		case SUPERSAMPLING_MODE_NORMAL: {
			settings->layerFlags |= XR_COMPOSITION_LAYER_SETTINGS_NORMAL_SUPER_SAMPLING_BIT_FB;
		} break;
		case SUPERSAMPLING_MODE_QUALITY: {
			settings->layerFlags |= XR_COMPOSITION_LAYER_SETTINGS_QUALITY_SUPER_SAMPLING_BIT_FB;
		} break;
}

switch ((SharpeningMode)(int)p_property_values.get(SHARPENING_MODE_PROPERTY_NAME, SHARPENING_MODE_NORMAL)) {
		case SHARPENING_MODE_NORMAL: {
			settings->layerFlags |= XR_COMPOSITION_LAYER_SETTINGS_NORMAL_SHARPENING_BIT_FB;
		} break;
		case SHARPENING_MODE_QUALITY: {
			settings->layerFlags |= XR_COMPOSITION_LAYER_SETTINGS_QUALITY_SHARPENING_BIT_FB;
		} break;
	};
```

Reply:
Ah good suggestion, much cleaner!